### ISSUE TITLE
Model-Layer

### ISSUE THREAD

**Issue by shiretzet** [fchollet/keras#3564]
Is there a way to build a model and train it, then plug it in to a new network?

What i'm trying to achieve is a layer, with lets say
```python 
[(batch, samples, channel, row, column), Model] 
```
input, where i iterate through the samples dim (Theano-scan or TimeDistributed is also o.k. for that matter) and for each 
```python
(batch, channel, row, column) 
```
i run the Model.

Today i specify the model in a costume layer where i need to play with the building and calling of each layer (something like the juggling in TimeDistributed).

I think it will be a powerful tool for keras if it's not already exist in someway

**Comment by fchollet** [comment]
Models are subclasses of Layers. So you can use a model as a layer (by
calling it on a tensor, following the functional API).

**Comment by shiretzet** [comment]
Ok, so let me give an example where there is a problem:
```python
import theano.tensor as T
import theano
from keras.layers.wrappers import Wrapper
from keras.layers import Dense, Dropout, Flatten, Input
from keras.models import Model

class MyWrapper(Wrapper):
    def __init__(self, model, **kwargs):
        self.layer = model
        super(MyWrapper, self).__init__(model, **kwargs)

    def build(self, input_shape=None):
        child_input_shape = (input_shape[0],)+input_shape[2:]
        if not self.layer.built:
            self.layer.build(input_shape)
            self.layer.built = True
        super(MyWrapper, self).build()

    def call(self, inputs, mask=None):
        batch_size = inputs.shape[0]
        img_height = inputs.shape[2]

        def _batch_itr(sample, img_batch):
            img_spl = img_batch[sample]

            def _model_itr(idx, img):
                img = img.dimshuffle('x',0,1,2)
                res = self.layer.call(img[:,:,idx:,:])
                return T.flatten(res)

            result, updates = theano.scan(fn=_model_itr, sequences=T.arange(img_height-1), non_sequences=img_spl)
            return result, updates

        output, updates = theano.scan(fn=_batch_itr, sequences=T.arange(batch_size), non_sequences=inputs)
        result = output[0]
        inner_updates = output[1]
        return result

    def get_output_shape_for(self, input_shape):
        return (input_shape[0],input_shape[2]-1)

### Inner model ###
inner_in = Input(shape=(3, 28, 28), name='inner_input_img')

inner = Flatten()(inner_in)
inner = Dense(1024, activation='relu')(inner)
inner = Dense(512, activation='relu')(inner)
inner = Dropout(0.5)(inner)
inner_out = Dense(1, activation='sigmoid', name='inner_result')(inner)

inner_model = Model(inner_in, inner_out)

### Outer model ###
outer_in = Input(shape=(3, 28, 28), name='input_img')

outer_wrp = MyWrapper(inner_model)(outer_in)
# shape :: (batch,27)
outer_out = Dense(1, activation='sigmoid', name='outer_result')(outer_wrp)

outer_model = Model(outer_in, outer_out)

outer_model.compile(optimizer='sgd',
                      loss={'outer_result': 'binary_crossentropy'},
                      metrics={'outer_result': 'accuracy'})
```

The problem/question is:
Because we use dropout in the inner model, the inner theano.scan returns a "must use" update param.
On the other hand, we already accumulated the updates of the inner_model to its self.update.
So.... should i ignore the returned updates from theano.scan (like i did), or should i add them to the outter_model self.updates (BTW Wrapper.build should allready get the inner_model self.updates)?

**Comment by shiretzet** [comment]
OK, solved.

if anyone wants to do something like this, the only thing you need to change is the `update` type of theano.scan like i did (Keras use `tuple`s as updates, not `dict`):
```python
import theano.tensor as T
import theano
from keras.layers.wrappers import Wrapper
from keras.layers import Dense, Dropout, Flatten, Input
from keras.models import Model
import numpy as np

class MyWrapper(Wrapper):
    def __init__(self, model, **kwargs):
        self.layer = model
        super(MyWrapper, self).__init__(model, **kwargs)

    def build(self, input_shape=None):
        child_input_shape = (input_shape[0],)+input_shape[2:]
        if not self.layer.built:
            self.layer.build(input_shape)
            self.layer.built = True
        super(MyWrapper, self).build()

    def call(self, inputs, mask=None):
        img_height = inputs.shape[2]

        def _batch_itr(img_batch):
            img_spl = img_batch

            def _model_itr(idx, img):
                img = img.dimshuffle('x',0,1,2)
                res = self.layer.call(img)
                return T.flatten(res)

            result, updates = theano.scan(fn=_model_itr, sequences=T.arange(img_height-1), non_sequences=img_spl)
            return T.flatten(result), updates

        output, updates = theano.scan(fn=_batch_itr, sequences=inputs)
        self.updates.append(tuple(updates.items())[0])
        return output

    def get_output_shape_for(self, input_shape):
        return (input_shape[0],input_shape[2]-1)

inner_in = Input(shape=(3, 28, 28), name='inner_input_img')

inner = Flatten()(inner_in)
inner = Dense(1024, activation='relu')(inner)
inner = Dense(512, activation='relu')(inner)
inner = Dropout(0.5)(inner)
inner_out = Dense(1, activation='sigmoid', name='inner_result')(inner)

inner_model = Model(inner_in, inner_out)

inner_model.compile(optimizer='sgd',
                      loss={'inner_result': 'binary_crossentropy'},
                      metrics={'inner_result': 'accuracy'})

outer_in = Input(shape=(3, 28, 28), name='input_img')

outer_wrp = MyWrapper(inner_model)(outer_in)
# shape :: (batch,27)
outer_out = Dense(1, input_shape=(27,), activation='sigmoid', name='outer_result')(outer_wrp)

outer_model = Model(outer_in, outer_out)

outer_model.compile(optimizer='sgd',
                      loss={'outer_result': 'binary_crossentropy'},
                      metrics={'outer_result': 'accuracy'})

X_train = np.arange(10*3*28*28).reshape((10,3,28,28))
Y_train = np.zeros((10,1))

Y_train[0,0] = 1
Y_train[5,0] = 1

print(X_train.shape)
print(Y_train.shape)

print("Fitting...")
outer_model.fit(X_train, Y_train,batch_size=1, nb_epoch=1, verbose=1)
```
